packages/expo-google-sign-in/ios/EXGoogleSignIn/EXGoogleSignIn.m

```diff
@@ -0,0 +1,250 @@
+                hostedDomain:options[@"hostedDomain"]
+                    serverId:options[@"webClientId"]
+                   loginHint:options[@"accountName"]
+                    language:options[@"language"]
+                 openIDRealm:options[@"openIdRealm"]];
+  resolve([NSNull null]);
+}
+
+EX_EXPORT_METHOD_AS(signInSilentlyAsync,
+                    signInSilentlyAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  if ([self.authTask update:@"signInSilentlyAsync" resolver:resolve rejecter:reject])
+    [[GIDSignIn sharedInstance] signInSilently];
+}
+
+EX_EXPORT_METHOD_AS(signInAsync,
+                    signInAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  if ([self.authTask update:@"signInAsync" resolver:resolve rejecter:reject])
+    [[GIDSignIn sharedInstance] signIn];
+}
+
+EX_EXPORT_METHOD_AS(signOutAsync,
+                    signOutAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  [[GIDSignIn sharedInstance] signOut];
+  resolve([NSNull null]);
+}
+
+EX_EXPORT_METHOD_AS(disconnectAsync,
+                    disconnectAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  if ([self.authTask update:@"disconnectAsync" resolver:resolve rejecter:reject])
+    [[GIDSignIn sharedInstance] disconnect];
+}
+
+EX_EXPORT_METHOD_AS(isSignedInAsync,
+                    isSignedInAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  resolve(@([[GIDSignIn sharedInstance] hasAuthInKeychain]));
+}
+
+EX_EXPORT_METHOD_AS(getCurrentUserAsync,
+                    getCurrentUserAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  resolve([EXGoogleSignIn jsonFromGIDGoogleUser:[[GIDSignIn sharedInstance] currentUser]]);
+}
+
+EX_EXPORT_METHOD_AS(getPhotoAsync,
+                    getPhotoAsync:(NSNumber *)size
+                    resolver:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  GIDGoogleUser *user = [[GIDSignIn sharedInstance] currentUser];
+  if (user && user.profile.hasImage) {
+    NSURL *imageURL = [user.profile imageURLWithDimension:[size unsignedIntegerValue]];
+    if (imageURL) {
+      resolve([imageURL absoluteString]);
+      return;
+    }
+  }
+  // TODO: Bacon: should we error out?
+  //  reject(E_EXCEPTION, @"Could not get photoURL", nil);
+  resolve([NSNull null]);
+}
+
+
+EX_EXPORT_METHOD_AS(getTokensAsync,
+                    getTokensAsync:(EXPromiseResolveBlock)resolve
+                    rejecter:(EXPromiseRejectBlock)reject)
+{
+  GIDGoogleUser *currentUser = [GIDSignIn sharedInstance].currentUser;
+  GIDAuthentication *auth = currentUser.authentication;
+  [auth getTokensWithHandler:^void(GIDAuthentication *authentication, NSError *error) {
```

Comment:
If the user is `nil`, won't this ever get called and promise never resolved/rejected? Are we sure `user` will always be there when this method is called?

Reply:
i think this is covered by the `getCurrentUserOrReject` selector, but this selector is a bit confusing and i'm not sure it's really necessary (probably only saving 1-2 lines of code each time it's called). @EvanBacon I'd probably just reject directly in each method to make your intent really clear.